zppy/__main__.py

```diff
@@ -73,47 +73,54 @@ def main():  # noqa: C901
             raise OSError("Cannot create script directory")
         pass
 
+    if config["default"]["machine"] == "":
+        # MachineInfo below will then call `discover_machine()`,
+        # which only works on log-in nodes.
+        machine = None
+    else:
+        # If `machine` is set, then MachineInfo can bypass the
+        # `discover_machine()` function.
+        machine = config["default"]["machine"]
+    machine_info = MachineInfo(machine=machine)
+    default_machine = machine_info.machine
+    (
+        default_account,
+        default_partition,
+        default_constraint,
+        _,
+    ) = machine_info.get_account_defaults()
+    unified_base = machine_info.config.get("e3sm_unified", "base_path")
+
     # Determine machine to decide which header files to use
-    tmp = os.getenv("HOSTNAME")
-    if tmp:
-        machine_info = MachineInfo()
-        unified_base = machine_info.config.get("e3sm_unified", "base_path")
-        if tmp.startswith("compy"):
-            machine = "compy"
-            environment_commands = (
-                f"source {unified_base}/load_latest_e3sm_unified_compy.sh"
-            )
-            account = "e3sm"
-        elif tmp.startswith("cori"):
-            machine = "cori"
-            partition = config["default"]["partition"]
-            if partition not in ["haswell", "knl"]:
-                raise ValueError(
-                    f'Expected Cori partition to be "haswell" or '
-                    f'"knl" but got: {partition}'
-                )
-            # This will be the [default] environment_commands
-            environment_commands = (
-                f"source {unified_base}/load_latest_e3sm_unified_cori-{partition}.sh"
-            )
-            account = "e3sm"
-        elif tmp.startswith("blues"):
-            machine = "anvil"
-            environment_commands = (
-                f"source {unified_base}/load_latest_e3sm_unified_anvil.sh"
-            )
-            account = "condo"
-        elif tmp.startswith("chr"):
-            machine = "chrysalis"
-            environment_commands = (
-                f"source {unified_base}/load_latest_e3sm_unified_chrysalis.sh"
+    if config["default"]["machine"] == "":
+        config["default"]["machine"] = default_machine
+    # Determine account
+    if config["default"]["account"] == "":
+        config["default"]["account"] = default_account
+    # Determine partition
+    if config["default"]["partition"] == "":
+        if config["default"]["machine"] == "cori":
+            config["default"]["partition"] = default_constraint
+        else:
+            config["default"]["partition"] = default_partition
```

Comment:
I believe this should be fixed in zppy.  Haswell/KNL are constraints, not partitions.

Reply:
Again, not worth fixing for a machine that's about to go.